Foundation/NSOperation.swift

```diff
@@ -71,82 +140,392 @@ public class NSOperation : NSObject {
+        didSet { updateUnderlyingQueue() }
+    }
+
+    unowned(unsafe) public var underlyingQueue: dispatch_queue_t { return _underlyingQueue._queue }
+
+    private lazy var _underlyingQueue: DispatchQueueBridge = {
+        return self.concurrentQueueWithName(self.name, qos: self.qualityOfService)
+    }()
+
+    private let _dispatchGroup = DispatchGroupBridge()
+    private var _concurrencyLimitingSemaphore: DispatchSemaphoreBridge
+
     public override init() {
-        NSUnimplemented()
+        _concurrencyLimitingSemaphore = DispatchSemaphoreBridge(count: maxConcurrentOperationCount)
     }
-    
-    public func addOperation(op: NSOperation) {
-        NSUnimplemented()
+
+    private init(queue: DispatchQueueBridge, name aName: String, maxConcurrentOperationCount max: Int = NSOperationQueueDefaultMaxConcurrentOperationCount) {
+        self.name = aName
+        self.maxConcurrentOperationCount = max
+
+        _concurrencyLimitingSemaphore = DispatchSemaphoreBridge(count: maxConcurrentOperationCount)
     }
 
-    public func addOperations(ops: [NSOperation], waitUntilFinished wait: Bool) {
-        NSUnimplemented()
+    public func addOperation(operation: NSOperation) {
+        addOperations([operation], waitUntilFinished: false)
     }
-    
-    public func addOperationWithBlock(block: () -> Void) {
-        NSUnimplemented()
+
+    public func addOperations(operations: [NSOperation], waitUntilFinished wait: Bool) {
+        _operationsLock.lock()
+        _operations += operations
+        _operationsLock.unlock()
+
+        _suspendedLock.lock()
+        defer {  _suspendedLock.unlock() }
+        if !suspended {
+            executeOperationsAndWait(wait)
+        }
     }
-    
+
+    private func executeOperations(operations: [NSOperation]) {
+        for operation in operations {
+            _dispatchGroup.enter()
+
+            _underlyingQueue.dispatchAsynchronously({
+                self._concurrencyLimitingSemaphore.wait(DispatchTimeForever)
+
+                self.executeOperation(operation)
+
+                self._concurrencyLimitingSemaphore.signal()
+                self._dispatchGroup.leave()
+            })
+        }
+    }
+
+    private func executeOperation(operation: NSOperation) {
+        if operation.asynchronous {
+            operation.start()
+            operation.waitUntilFinished()
+        } else {
+            operation.main()
+        }
+    }
+
+    private func operationsReadyToExecute(ops: [NSOperation]) -> [NSOperation] {
+        let readyOperations = ops.filter({ operation in operation.ready })
+        let readyAndSortedOperations = readyOperations.sort({ operation1, operation2 in
+            operation1.queuePriority.rawValue > operation2.queuePriority.rawValue
+        })
+
+        return readyAndSortedOperations
+    }
+
+    private func executeOperationsAndWait(wait: Bool) {
```

Comment:
this should actually just enqueue blocks to a dispatch queue; the behavior listed will incur an overhead to require re-calling the execution upon finishing an operation.